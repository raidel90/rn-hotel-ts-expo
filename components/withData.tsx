import React from "react";
import { ErrorScreen } from "./ErrorScreen";
import { LoadingScreen } from "./LoadingScreen";

export type WithDataProps<TData, TProps> = { data: TData } & TProps;

export interface UseDataResult<TData> {
  error: unknown;
  data: TData | undefined;
}

export function withData<TData>(useData: () => UseDataResult<TData>) {
  return function <TProps>(WrappedComponent: React.ComponentType<{ data: TData } & TProps>) {
    return function (props: TProps) {
      const { error, data } = useData();

      if (error) {
        return <ErrorScreen error={error} />;
      }

      if (!data) {
        return <LoadingScreen />;
      }

      return (
        <WrappedComponent {...props} data={data} />
      );
    }
  }
}