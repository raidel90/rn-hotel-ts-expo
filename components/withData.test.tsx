import { UseDataResult, withData, WithDataProps } from "./withData";
import { Text } from "react-native";
import { createRenderer } from "react-test-renderer/shallow";

const renderer = createRenderer();

describe("withData", () => {
  it("should display a loading screen when data is undefined", () => {
    
    type WrappedProps = WithDataProps<string, unknown>;
    const WrappedComponent = ({ data }: WrappedProps) => <Text>{data}</Text>;

    const useData = (): UseDataResult<string> => ({
      data: undefined,
      error: undefined
    });

    const Component = withData(useData)(WrappedComponent);

    const result = renderer.render(<Component />);
    expect(result).toMatchSnapshot();
    
  });
});