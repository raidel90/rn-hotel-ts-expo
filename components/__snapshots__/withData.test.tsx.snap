// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withData should display a loading screen when data is undefined 1`] = `<LoadingScreen />`;
